front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx

```diff
@@ -38,83 +35,122 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
-        id: formatPacedTrainIdToExceptionId(id, exception.key),
-        // TODO exceptions : find a better way to generate added exception names
-        trainName: computeOccurrenceName(name, occurrencesCount + i),
-        rollingStock,
-        // An added exception will always have a least a start time in its exceptions
-        startTime: new Date(exception.start_time!.value),
-        stopsCount,
-        exceptionChangeGroups: omit(exception, ['key']),
-        ...(isValid
-          ? {
-              isValid: true,
-              // TODO exceptions : update the arrival time if the exception is in the paced train summaries
-              arrivalTime: new Date(Date.parse(exception.start_time!.value) + duration!.ms),
-              pathLength,
-              mechanicalEnergyConsumed,
-              duration,
-            }
-          : {
-              isValid: false,
-            }),
+  const [occurrences, setOccurrences] = useState<Occurrence[]>([]);
+  const occurrencesCount = getOccurrencesNb(paced);
+
+  const [getRollingStockByName] =
+    osrdEditoastApi.endpoints.getRollingStockNameByRollingStockName.useLazyQuery();
+
+  useEffect(() => {
+    const buildOccurrences = async () => {
+      const computedOccurrences: Occurrence[] = [];
+
+      // Handle indexed occurrences
+      for (let i = 0; i < occurrencesCount; i += 1) {
+        const occurrenceId = formatPacedTrainIdToIndexedOccurrenceId(id, i);
+        const occurrenceStartTime = dayjs(startTime)
+          .add(i * paced.interval.ms, 'ms')
+          .toDate();
+        const occurrenceArrivalTime = dayjs(arrivalTime)
+          .add(i * paced.interval.ms, 'ms')
+          .toDate();
+
+        const correspondingException = findExceptionWithOccurrenceId(exceptions, occurrenceId);
+
+        let occurrenceRollingStock = rollingStock;
+        if (correspondingException?.rolling_stock) {
+          const promisedRollingStock = getRollingStockByName({
+            rollingStockName: correspondingException.rolling_stock.rolling_stock_name,
+          });
+          occurrenceRollingStock = await promisedRollingStock.unwrap();
+          // we don't want to subscribe to the endpoint to prevent unnecessary calls
+          promisedRollingStock.unsubscribe();
+        }
+
+        computedOccurrences.push({
+          id: occurrenceId,
+          trainName: correspondingException?.train_name?.value ?? computeOccurrenceName(name, i),
+          rollingStock: occurrenceRollingStock,
+          startTime: correspondingException?.start_time?.value
+            ? new Date(correspondingException?.start_time?.value)
+            : occurrenceStartTime,
+          stopsCount: correspondingException?.path_and_schedule
+            ? (correspondingException?.path_and_schedule.schedule?.filter(
+                (step) => step.stop_for && Duration.parse(step.stop_for).ms > 0
+              ).length ?? 0) + 1 // +1 to take the final stop (destination) into account,
```

Comment:
- thank you for the comment, it's clearer with it
- even if the duration of the stop is 0 ms, the train will stop so I think we should consider it as a stop. Maybe we can remove the second part of the test ?
```suggestion
            ? (correspondingException.path_and_schedule.schedule.filter((step) => step.stop_for)
                .length ?? 0) + 1 // +1 to take the final stop (destination) into account,
```

Reply:
Hmm if we remove this condition, the destination will be counted twice most of the time as we instantiate it with a `stopFor: 0` so we need to remove the `+ 1` at the end. I think this 0 stop is a hack we use to make the train stop at the end of its itinerary but making a stop: 0 in the middle of the itinerary doesn't make sense to me irl ?

Reply:
Done